Add `--reload` flag to the `serve run` subcommand

volks73 opened the pull request:
<!-- Thank you for your contribution! Please review https://github.com/ray-project/ray/blob/master/CONTRIBUTING.rst before opening a pull request. -->

<!-- Please add a reviewer to the assignee section when you create a PR. If you don't have the access to it, we will shortly find a reviewer and assign them to your PR. -->

## Why are these changes needed?

I have implemented a custom watch using the `watchfiles` package, formerly named `watchgod`, for my own local development of Ray Serve applications. It would be nice if this functionality existed in the `serve` Command Line Interface (CLI). I have searched the GitHub issues and Ray forums. This feature has been requested off and on several times.

This PR adds the `--reload` flag to the `serve run` subcommand. It will listen for file changes in the specified working directory and redeploy the application. This is similar uvicorn and other web frameworks. It is useful for application development.

The `watchfiles` package is added to the `requirements.txt` file and the `setup.py` file as indicated.

## Related issue number

- #26341
- #24790
- #16094
- #15505

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [x] This PR is not tested :(

I have tested by creating an `example` folder in the project root, `mkdir example && example`, then creating the following example Ray Serve application, `example_serve.py`, based on the Key Concepts documentation:

```python
from ray import serve

@serve.deployment
class MyFirstDeployment:
    def __init__(self, msg):
        self.msg = msg

    def __call__(self):
        return self.msg


my_first_app = MyFirstDeployment.bind("Hello World!")
```

I started the Ray Serve application with following command from within the `example` folder:

```sh
serve run --reload --working-dir=./ example_serve:my_first_app
```

This yields the following output:

```sh
2023-08-13 17:44:49,851 INFO scripts.py:420 -- Running import path: 'example_serve:my_first_app'.
2023-08-13 17:44:55,768 INFO worker.py:1612 -- Started a local Ray instance. View the dashboard at http://127.0.0.1:8265
2023-08-13 17:44:55,772 INFO packaging.py:518 -- Creating a file package for local directory './'.
2023-08-13 17:44:55,773 INFO packaging.py:346 -- Pushing file package 'gcs://_ray_pkg_579ed46ebfd67af1.zip' (0.00MiB) to Ray cluster...
2023-08-13 17:44:55,773 INFO packaging.py:359 -- Successfully pushed file package 'gcs://_ray_pkg_579ed46ebfd67af1.zip'.
(ServeController pid=58220) INFO 2023-08-13 17:44:58,267 controller 58220 controller.py:357 - Finished recovering deployments after 0.00033402442932128906s.
(HTTPProxyActor pid=58249) INFO 2023-08-13 17:44:59,445 http_proxy 192.168.1.57 http_proxy.py:904 - Proxy actor 8f51acf84b79d0e8004b4c9901000000 starting on node 3f223d4ed49dce22ca1474482841f62b3e232754755644273ba0750f.
(ServeController pid=58220) INFO 2023-08-13 17:44:59,501 controller 58220 deployment_state.py:1358 - Deploying new version of deployment default_MyFirstDeployment.
(HTTPProxyActor pid=58249) INFO:     Started server process [58249]
(ServeController pid=58220) INFO 2023-08-13 17:44:59,604 controller 58220 deployment_state.py:1628 - Adding 1 replica to deployment default_MyFirstDeployment.
2023-08-13 17:45:01,512 SUCC scripts.py:461 -- Deployed Serve app successfully.
```

Then, I change the "Hello World" to "Hello Me", and I get the following output:

```sh
2023-08-13 17:45:13,460 INFO scripts.py:476 -- Files changed. Redeploying Serve app.
(ServeController pid=58220) INFO 2023-08-13 17:45:13,503 controller 58220 deployment_state.py:1358 - Deploying new version of deployment default_MyFirstDeployment.
(ServeController pid=58220) INFO 2023-08-13 17:45:13,614 controller 58220 deployment_state.py:1511 - Stopping 1 replicas of deployment 'default_MyFirstDeployment' with outdated versions.
(ServeController pid=58220) INFO 2023-08-13 17:45:15,942 controller 58220 deployment_state.py:1628 - Adding 1 replica to deployment default_MyFirstDeployment.
^C2023-08-13 17:45:57,611       INFO scripts.py:485 -- Got KeyboardInterrupt, shutting down...
(ServeController pid=58220) INFO 2023-08-13 17:45:57,692 controller 58220 deployment_state.py:1653 - Removing 1 replica from deployment 'default_MyFirstDeployment'.
```

edoakes commented on python/ray/serve/scripts.py:
```diff
@@ -323,6 +324,16 @@ def deploy(config_file_name: str, address: str):
+        "Listens for changes to files in the working directory, --working-dir "
+        "or the working_dir in the --runtime-env, and automatically redeploys "
+        "the application. This will block until Ctrl-C'd, then clean up the "
+        "app. The --blocking/--non-blocking flag is ignored."
```
Instead of ignoring, can we raise an explicit exception when `--non-blocking` is provided?

volks73 replied:
I have added raising a `click.ClickException` if `reload` is True and `blocking` is False.

The message is:

>The --non-blocking option conflicts with the --reload option.

I am open to different wording.

edoakes commented on python/ray/serve/scripts.py:
```diff
@@ -323,6 +324,16 @@ def deploy(config_file_name: str, address: str):
     ),
 )
+@click.option(
+    "--reload",
```
let's add a shorter alias for convenience, `"-r"`

volks73 replied:
I have added the shorter alias.

edoakes commented on python/ray/serve/scripts.py:
```diff
@@ -454,6 +466,16 @@ def run(
+                    yield_on_timeout=True,
+                ):
+                    if changes:
+                        cli_logger.info("Files changed. Redeploying Serve app.")
```
```suggestion
                        cli_logger.info(f"Detected file change in path {working_dir}. Redeploying Serve app.")
```
helps make the behavior a little more clear

edoakes commented on python/ray/serve/scripts.py:
```diff
@@ -454,6 +466,16 @@ def run(
         else:
+            if reload:
+                for changes in watchfiles.watch(
+                    working_dir,
```
what happens if `working_dir` is not provided? for local development, maybe let's default to `"."` and/or pick up the `--app-dir` if it's specified?

volks73 replied:
I have added to defaulting to `./` if the working directory is not specified.

edoakes reviewed (commented):
@volks73 thanks so much for adding this, it'll be a great UX improvement!

I left a few small comments on naming/behavior.

In addition, could you please add a test case to `python/ray/serve/tests/test_cli.py`? Should be of the form: start background process running `serve run` with reload, check app is deployed and returns initial response, update file in the working dir to return a new response, `wait_for_condition` that the app returns the new response.

volks73 commented:
@edoakes I have signed the previous and current commits. Sorry about that.

I have added a test to the `python/ray/serve/tests/test_cli_2.py` file instead of the `test_cli.py` file because I saw all of the `serve run` tests are in the `test_cli_2.py` file. I am not sure if the test is correct because I added a fixture to use a temporary directory for the working directory and change to that directory.

edoakes commented on python/ray/serve/scripts.py:
```diff
@@ -454,6 +467,26 @@ def run(
+                ):
+                    if changes:
+                        cli_logger.info(
+                            f"Detected file change in path {working_dir}. Redeploying Serve app."
```
```suggestion
                            f"Detected file change in path {watch_dir}. Redeploying Serve app."
```

edoakes commented on python/ray/serve/scripts.py:
```diff
@@ -454,6 +467,26 @@ def run(
+            if reload:
+                if working_dir:
+                    watch_dir = working_dir
+                else:
+                    watch_dir = "./"
```
```suggestion
                else:
                    watch_dir = app_dir
```
`app_dir` defaults to `"."` and is the directory that the import_path is imported from

edoakes commented on python/ray/serve/scripts.py:
```diff
@@ -454,6 +467,26 @@ def run(
             visualizer = GraphVisualizer()
             visualizer.visualize_with_gradio(handle)
         else:
+            if reload and not blocking:
+                raise click.ClickException(
+                    "The --non-blocking option conflicts with the --reload option."
+                )
+            if reload:
```
```suggestion
        elif reload:
            if blocking:
                raise click.ClickException(
                    "The --non-blocking option conflicts with the --reload option."
                )
            # unindent the reload codepath
```

edoakes reviewed (commented):
Test looks great, thanks! Just some small nits then LGTM.

edoakes commented on python/ray/serve/tests/test_cli_2.py:
```diff
@@ -822,5 +822,55 @@ def test_deployment_contains_utils(ray_start_stop):
+            "reload_serve:msg_app",
+        ]
+    )
+    wait_for_condition(lambda: ping_endpoint("Message") == "Hello World!", timeout=10)
```
I believe the test is failing because this should be:
```suggestion
    wait_for_condition(lambda: ping_endpoint("") == "Hello World!", timeout=10)
```
(by default it's hosted at `"http://localhost:8000/"` with no route prefix)

edoakes reviewed (commented):
Test case is failing, pointed out what looks like the issue:
https://buildkite.com/ray-project/oss-ci-build-pr/builds/31759#0189f509-1af0-4fa7-be7c-320a541be7ac/6-2237

volks73 commented:
@edoakes Thanks for the suggestions. I have updated the code and possibly fixed the test. I forgot to add an import statement to the "test" deployment that is created. This was causing an import error further up in the test run. Hopefully with this fix and your suggestion for pinging the endpoints, the test will pass.

edoakes commented:
FYI apparently github is having a minor outage at the moment, so you may need to re-kick off the build later today once it's resolved

volks73 commented:
@edoakes I have kicked off the rebuild by merging the latest changes from `master`.

edoakes commented:
Test appears to still be failing. Let me see if I can fix it:
https://buildkite.com/ray-project/oss-ci-build-pr/builds/32159#0189fee5-4283-486e-aa55-8d8333474a00/6-2248

edoakes reviewed (approved):
![image](https://github.com/ray-project/ray/assets/9871461/bb209899-c1d2-4134-91f6-4547a28944d5)

volks73 commented:
@edoakes Great! Thank you for fixing the test and the app reloading. I don't know of another solution than sleeping for a while either.

edoakes commented:
Doc build failed due to missing dependency, hopefully fixed 🤞

volks73 commented:
@edoakes I am having trouble following the tests. Did updating the dependencies for the docs work?

edoakes commented on python/setup.py:
```diff
@@ -336,6 +336,7 @@ def get_packages(self):
         "aiosignal",
         "frozenlist",
         "requests",
+        "watchfiles",
```
@ericl @richardliaw can one of you stamp for this please? I think mine doesn't count because I pushed a commit to the branch.

richardliaw replied:
should this be in `serve` instead of a core dependency?

richardliaw replied:
https://github.com/ray-project/ray/pull/38389/files#diff-eb8b42d9346d0a5d371facf21a8bfa2d16fb49e213ae7c21f03863accebe0fcfL272 should go here instead

volks73 replied:
I have moved the `watchfiles` dependency to the `serve` dependencies.

edoakes commented:
@volks73 all the relevant tests are passing (we have some typical flakiness in CI). Just need one more approval and should get merged in time for 2.7 branch cut!

richardliaw commented on python/requirements.txt:
```diff
@@ -17,6 +17,7 @@ pyyaml
 aiosignal
 frozenlist
 requests
+watchfiles
```
should we pin this?

edoakes replied:
no need to pin it, we are using a public/stable interface

richardliaw reviewed (commented):
move the dependency to `serve` instead of `default`?

volks73 commented:
@edoakes I have moved the `watchfiles` dependency to the `serve` dependencies as requested.

richardliaw reviewed (approved):
approving from docs side

edoakes commented:
@volks73 merged 🎉 thanks again for contributing. This has been on the backlog for a long time and happy to see it done :)

volks73 commented:
@edoakes Your welcome and thank you for helping with the PR and overall for the Ray project.
